docs/design/jet/025-jet-job-member-pruning.md

```diff
@@ -0,0 +1,197 @@
+|Technical Reviewers|TBD|
+|Version|5.4|
+
+## Background
+
+Before Hazelcast Platform 5.4, Jet job was always deployed to all data members (specifically - all not-lite members). 
+If some DAG vertex isn’t using all members, it creates no-op processors on the rest, but Jet still creates queues 
+to/from those vertices and starts the processors, even though it completes immediately and the queues are closed with a `DONE_ITEM`. 
+If some member isn’t used at all, the DAG is still deployed to it, and the master has to send it to it and wait for 
+the completion. Even though the processors are no-op, or have no data to process, it’s an unnecessary overhead, 
+which becomes noticeable in very small batch jobs.
+
+## Terminology
+
+- Member pruning - prevention of cluster members without requested data on board to be involved in job execution;
+- Processor pruning - elimination of redundant (it can work only with input) stage processor creation;
+- (IMap) Partition pruning - extract partition condition and assigning only required partition to read by `ReadMapOrCacheP`;
+
+## Goals
+
+The goals of that initiative is corresponding with items enumerated in 'Terminology' section : 
+
+- deploy a job only on members with required partitions.
+- prune processors which are not required for job execution - e.g, their input processor doesn't produce data and 
+the processor itself can work only with input.
+- limit count partitions involved in IMap scan.
+
+Non-goals are :
+- support any kind of pruning for streaming jobs. It may be considered to do later, but now it is not a case.
+- support migration-tolerance for member and processor pruning. 
+
+
+## Technical Design
+
+Let's split member pruning, processor pruning and partition pruning as separate stages, which can complement each other, 
+but on their own, each kind of pruning is still a good optimization.
+
+### Member pruning
+
+Member pruning by its own, may have pretty simple solution. To implement it, we need to support that solution 
+in `ExecutionPlanBuilder` (execution plan creation phase).  Previous `ProcessorMetaSupplier#get` contract was a reason 
+for strict assumptions in execution plan creation algorithm, but that has changed.
+
+#### Use cases for member pruning
+
+**Important note**: by default, the basic unit on example pictures will be DAG vertices, but sometimes we will 
+use processors as basic unit instead. We will explicitly declare it.
+
+##### Member pruning for simple map scan
+
+Prune all members except Member 1 for scanning job. We want to highlight this simple case, because we assume that 
+the bigger part of the submitted queries looks like `SELECT * FROM map WHERE ...`. So, if you may hint the optimizer
+with partitionKey, the submitted job would be even local.
+
+```
+- SCAN[map2, partitionKey=1]
+```
+
+##### Member pruning for Scan -> Hash Join in big cluster
+
+Prune members without any connections, so, only Member 1 and Member K left in cluster.
```

Comment:
how do you determine which members have _connections_? Connection is not defined and is not obvious. It is not an edge in the DAG